csrc/kernel_cache.h

```diff
@@ -323,57 +343,71 @@ class TORCH_CUDA_CU_API InputsIdLookup : public NonCopyable {
-//! shape, strides, type e.t.c.. However, the contract right here is fuzzy.
-//! Different executor applies their own protocol of what is a unique
-//! computational graph. e.g. Legacy Executor embeds tensor type &
-//! dimensionality in the graph, while Profiling Executor keeps symbolic shape
-//! as well as stride order in the graph as well.
+//! The content of input tensors does not affect the structure of the Fusion
+//! graph or the validity of compiled CUDA kernels, so we know that if all input
+//! shapes are the same, then a concrete Fusion which is valid for one set of
+//! inputs will be valid for another. We leverage this fact to reduce latency
+//! when Fusions are called repeatedly with inputs that differ only in their
+//! tensor content. Given inputs, we first compute an ID using
+//! InputsIdLookup::lookupId that encodes the shapes of TensorView inputs, along
+//! with values of any integer-valued input scalars that might affect
+//! concretization. This ID is guaranteed not to conflict unless the inputs can
+//! be executed by the same compiled Fusion. It is mapped to a segmented and
+//! compiled FusionKernelRuntime that can be immediately run. This is the most
+//! common, lowest-latency path and is followed when the Fusion is repeatedly
+//! evaluated with same-shaped inputs.
 //!
-//! Our definition of a "unique" computational graph is aligned with `Fusion`
-//! IR, hence the requirement extends to meta information on input tensors.
-//! Which means, for each input tensor, following properties are fixed:
-//!     a) stride order;
-//!     b) contiguity information;
-//!     c) broadcasting semantics (size-1 or not);
-//!     d) rank;
-//!     e) scalar type;
+//! When there is no FusionKernelRuntime matching the given input ID, we first
+//! map the inputs to a concrete Fusion. For static Fusions this is trivial,
+//! but when the Fusion is dynamic it means we must use the inputs to
+//! "concretize" the Fusion by performing replacements such that the Fusion no
+//! longer contains Symbolic IterDomains. A static Fusion is considered to be
+//! already concretized. As discussed above, each concretized Fusion might have
+//! multiple FusionKernelRuntimes applying to different ranges of input shapes.
+//! In the second layer of post-definition caching, we map concretized Fusions
+//! (in the form of a DynamicTransformConcretizationInfo object) to a vector of
+//! FusionKernelRuntimes, and check whether each one is able to run the present
+//! inputs. If not, we create a new FusionKernelRuntime for those inputs and
+//! record it in the list (as well as recording a mapping from the input ID to
+//! the new runtime).
 //!
+//! In the case of a dynamic Fusion, input scalars such as integer parameters to
+//! a model could potentially affect the structure of a concretized Fusion, so
+//! we take care to include those scalars in the input ID along with the extents
+//! of tensor arguments.
```

Comment:
^ Updated note on "post-definition" caching

Reply:
Ah of course. I was focusing on `Val`s and forgot to replace that. Will update.

Reply:
Updated to place the original note back in place, and also updated my new note for correctness about additional tensor properties other than shape.